### PR TITLE
search onboarding tour redesign

### LINKED ISSUE #20801: "new: search contexts" popup should only appear if users have completed the search tour
image

### PR DESCRIPTION
**Pull request by novoselrok**
https://sourcegraph.slack.com/archives/C89KCDK5J/p1620914040040500

Fixes #20801.

### CODE REVIEW

**Review by lguychard — APPROVED** [review]
Code changes LGTM! Happy to see the `inputLocation` stuff go away 😄

**Review by limitedmage — APPROVED** [review]
Looks good!!

I'm seeing some odd minor glitches with this though:

image

- Contexts tour still shows along with the onboarding tour (I think is a separate issue?)
- Font color for contexts tour is incorrect in dark mode
- Very minor: Zooming in and scrolling horizontally has a weird effect (maybe this has always happened)
- Very minor: I'm concerned the new arrow design won't work for other types of tour popups, specifically #21049

**Comment by rrhyne** [comment]
> Very minor: I'm concerned the new arrow design won't work for other types of tour popups, specifically 

I wasn't aware the same tour functionality was used for monitors and contexts messaging. I would have expected a popover using the styles from without the weight of the tour code, so I hadn't planned on making this all work together in this iteration. 

Can we use the old white popover for the context and monitor element, but these styles for the tour?

**Comment by lguychard** [comment]
> I'm not as excited to see the input location not be consistent with the cursor as Loic is to see the code go ;)

Note: `inputLocation` in the code I was referring to pertained to whether the search input was rendered on the homepage or on the results page -- not to _cursor location in the input field_, which I think you are referring to here @rrhyne

**Comment by novoselrok** [comment]
> I'm not as excited to see the input location not be consistent with the cursor as Loic is to see the code go ;). I'm afraid this might have been a miscommunication with some of the other concepts in the design file.
> 
> Is there a technical reason it was removed?



> I'm not as excited to see the input location not be consistent with the cursor as Loic is to see the code go ;). I'm afraid this might have been a miscommunication with some of the other concepts in the design file.
> 
> Is there a technical reason it was removed?

Sorry to disappoint you, but the tour was never consistent with the input location. It was always positioned in the middle of the input. See current tour:
Screenshot 2021-05-19 at 10 28 53

I don't think the tour library we're using has the necessary features to implement this.

File: client/web/src/search/input/SearchBox.story.tsx [hunk 1]
```diff
@@ -44,6 +44,7 @@ const defaultProps: SearchBoxProps = {
     onSubmit: () => {},
     fetchAutoDefinedSearchContexts: mockFetchAutoDefinedSearchContexts(),
     fetchSearchContexts: mockFetchSearchContexts,
+    isSearchOnboardingTourActive: false,
```
[the remark this answers is not in the document]
Comment (novoselrok) [reply client/web/src/search/input/SearchBox.story.tsx:47]: Tried it, would take a bunch more refactoring and extracting props, and this PR is getting too complicated as it is. Can be done in the future, if we plan to extend the tour.

**Comment by rrhyne** [comment]
So sorry gents, I thought we did align to input location!!! Too much time in Figma I guess!!!

**Review by rrhyne — APPROVED** [review]
Looks great, thanks for the alterations!

**Review by limitedmage — COMMENTED** [review]
With the latest changes, the tour never seems to be completed and keeps showing up. I can dismiss it with the `x`. However I can't get the search contexts "tour" to appear again.

**Comment by novoselrok** [comment]
> With the latest changes, the tour never seems to be completed and keeps showing up. I can dismiss it with the `x`. However I can't get the search contexts "tour" to appear again.

That's really weird, do you have any specific reproduction steps? I made a recording of going through the tour and then re-enabling it once finished. You can also try running `localStorage.clear()` in your console to get a clean slate.

https://user-images.githubusercontent.com/6417322/118924552-8cb8d880-b93d-11eb-8d9e-c546ecc48b34.mov

**Comment by rrhyne** [comment]
Let's try leaving the tour open until the user closes it (since it only appears on the home page). If this tests bad in user testing, we can turn it off after the first use.

**Comment by novoselrok** [comment]
@rrhyne The onboarding tour is now dismissed only if the user cancels it.
@limitedmage _I think_ I found the bug you were talking about, it should be fixed now.

**Review by limitedmage — APPROVED** [review]
Looking and working great!!

**Comment by limitedmage** [comment]
Actually I have one more minor piece of feedback here: the last step should say `enter` instead of `return` if the user is on Windows or Linux, as that is what the key is called.

**Pull request merged**